test_rmw_implementation/test/test_qos_profile_check_compatible.cpp

```diff
@@ -0,0 +1,78 @@
+
+#ifdef RMW_IMPLEMENTATION
+# define CLASSNAME_(NAME, SUFFIX) NAME ## __ ## SUFFIX
+# define CLASSNAME(NAME, SUFFIX) CLASSNAME_(NAME, SUFFIX)
+#else
+# define CLASSNAME(NAME, SUFFIX) NAME
+#endif
+
+TEST(CLASSNAME(TestQoSProfilesAreCompatible, RMW_IMPLEMENTATION), compatible) {
+  rmw_ret_t ret;
+  rmw_qos_compatibility_type_t compatible;
+  // All of the provided profiles should be compatible with themselves
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_sensor_data, rmw_qos_profile_sensor_data, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_OK);
+  EXPECT_EQ(compatible, RMW_QOS_COMPATIBILITY_OK);
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_default, rmw_qos_profile_default, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_OK);
+  EXPECT_EQ(compatible, RMW_QOS_COMPATIBILITY_OK);
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_parameters, rmw_qos_profile_parameters, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_OK);
+  EXPECT_EQ(compatible, RMW_QOS_COMPATIBILITY_OK);
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_parameter_events, rmw_qos_profile_parameter_events, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_OK);
+  EXPECT_EQ(compatible, RMW_QOS_COMPATIBILITY_OK);
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_services_default, rmw_qos_profile_services_default, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_OK);
+  EXPECT_EQ(compatible, RMW_QOS_COMPATIBILITY_OK);
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_system_default, rmw_qos_profile_system_default, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_OK);
+  EXPECT_EQ(compatible, RMW_QOS_COMPATIBILITY_OK);
+}
+
+TEST(CLASSNAME(TestQoSProfilesAreCompatible, RMW_IMPLEMENTATION), error_on_unknown) {
+  rmw_ret_t ret;
+  rmw_qos_compatibility_type_t compatible;
+  // Expect an error code if a policy is "unknown"
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_sensor_data, rmw_qos_profile_unknown, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_INVALID_ARGUMENT);
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_default, rmw_qos_profile_unknown, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_INVALID_ARGUMENT);
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_parameters, rmw_qos_profile_unknown, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_INVALID_ARGUMENT);
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_parameter_events, rmw_qos_profile_unknown, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_INVALID_ARGUMENT);
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_services_default, rmw_qos_profile_unknown, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_INVALID_ARGUMENT);
+  ret = rmw_qos_profile_check_compatible(
+    rmw_qos_profile_system_default, rmw_qos_profile_unknown, &compatible, nullptr, 0u);
+  EXPECT_EQ(ret, RMW_RET_INVALID_ARGUMENT);
+}
```

Comment:
I assume the purpose of the tests here are to ensure that the function is vectored properly to its implementation and that the full test suite for QoS compatibility is over in rmw_dds_common.

Reply:
Yes. These tests also are run for every RMW. This is the reason that the tests are not as comprehensive as the tests in `rmw_dds_common`. I don't think we can make many assumptions about QoS compatibility if we consider an non-DDS RMW.